File: Go/ExampleTests/common/discounting_typed.go
package common

type DiscountingTyped struct {
	TotalPrice Dollar
	Discount Percentage
	Notes string
}

func NewDiscountingTypedFromString(s DiscountingString) DiscountingTyped {
	t := DiscountingTyped{}
	t.TotalPrice = Dollar(s.TotalPrice)
	t.Discount = Percentage(s.Discount)
	t.Notes = s.Notes
	return t
}

// ToJSONValue renders the struct as a plain map for encoding/json.
func (t DiscountingTyped) ToJSONValue() map[string]interface{} {
	return map[string]interface{}{
		"total_price": string(t.TotalPrice),
		"discount": string(t.Discount),
		"notes": t.Notes,
	}
}

func (t DiscountingTyped) ToJSON() (string, error) {
	return JSONWrite(t.ToJSONValue())
}

func NewDiscountingTypedFromJSONValue(m map[string]interface{}) (DiscountingTyped, error) {
	t := DiscountingTyped{}
	rawTotalPrice, err := JSONRequire(m, "total_price")
	if err != nil {
		return t, err
	}
	valTotalPrice, err := JSONAsString(rawTotalPrice, "total_price")
	if err != nil {
		return t, err
	}
	t.TotalPrice = Dollar(valTotalPrice)
	rawDiscount, err := JSONRequire(m, "discount")
	if err != nil {
		return t, err
	}
	valDiscount, err := JSONAsString(rawDiscount, "discount")
	if err != nil {
		return t, err
	}
	t.Discount = Percentage(valDiscount)
	rawNotes, err := JSONRequire(m, "notes")
	if err != nil {
		return t, err
	}
	valNotes, err := JSONAsString(rawNotes, "notes")
	if err != nil {
		return t, err
	}
	t.Notes = valNotes
	return t, nil
}

func NewDiscountingTypedFromJSON(text string) (DiscountingTyped, error) {
	m, err := JSONParseObject(text)
	if err != nil {
		return DiscountingTyped{}, err
	}
	return NewDiscountingTypedFromJSONValue(m)
}

func DiscountingTypedToJSONList(list []DiscountingTyped) (string, error) {
	arr := make([]interface{}, 0, len(list))
	for _, item := range list {
		arr = append(arr, item.ToJSONValue())
	}
	return JSONWrite(arr)
}

func DiscountingTypedFromJSONList(text string) ([]DiscountingTyped, error) {
	raw, err := JSONParseArray(text)
	if err != nil {
		return nil, err
	}
	result := make([]DiscountingTyped, 0, len(raw))
	for _, e := range raw {
		m, err := JSONAsObject(e, "DiscountingTyped")
		if err != nil {
			return nil, err
		}
		item, err := NewDiscountingTypedFromJSONValue(m)
		if err != nil {
			return nil, err
		}
		result = append(result, item)
	}
	return result, nil
}

File: Swift/ExampleTests/common/DiscountingTyped.swift
public struct DiscountingTyped {
    public let totalPrice: Dollar
    public let discount: Percentage
    public let notes: String

    public init(totalPrice: Dollar, discount: Percentage, notes: String) {
        self.totalPrice = totalPrice
        self.discount = discount
        self.notes = notes
    }

    public init(from s: DiscountingString) {
        self.totalPrice = Dollar(s.totalPrice)
        self.discount = Percentage(s.discount)
        self.notes = s.notes
    }

    public func toJSONValue() -> [String: Any] {
        return [
            "totalPrice": String(describing: totalPrice),
            "discount": String(describing: discount),
            "notes": notes,
        ]
    }

    public func toJSON() throws -> String {
        return try Json.write(toJSONValue())
    }

    public init(fromJSONValue m: [String: Any]) throws {
        self.totalPrice = Dollar(try Json.asString(Json.require(m, "totalPrice"), "totalPrice"))
        self.discount = Percentage(try Json.asString(Json.require(m, "discount"), "discount"))
        self.notes = try Json.asString(Json.require(m, "notes"), "notes")
    }

    public init(fromJSON text: String) throws {
        try self.init(fromJSONValue: Json.parseObject(text))
    }

    public static func toJSONList(_ list: [DiscountingTyped]) throws -> String {
        return try Json.write(list.map { $0.toJSONValue() })
    }

    public static func fromJSONList(_ text: String) throws -> [DiscountingTyped] {
        return try Json.parseArray(text).map {
            try DiscountingTyped(fromJSONValue: Json.asObject($0, "DiscountingTyped"))
        }
    }
}
